File: backend/routers/usage_trends.py
"""
usage_trends.py — Usage trends route.
Port of backend_legacy/routes/usageTrends.js.
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy import text
from sqlalchemy.engine import Connection

from ..dependencies import AuthUser, get_db, require_auth
from ..queries.analytics_shared import (
    VALID_METRICS,
    build_access_filter,
    get_metric_query,
    get_trend_insights,
)

router = APIRouter(prefix="/api/usage-trends", tags=["usage-trends"])

VALID_GRANULARITIES = {"day", "week", "month", "quarter"}


@router.get("/")
def usage_trends(
    granularity: str = Query("month"),
    metric: str = Query("uploaded_count"),
    auth: AuthUser = Depends(require_auth),
    conn: Connection = Depends(get_db),
):
    if granularity not in VALID_GRANULARITIES:
        granularity = "month"
    if metric not in VALID_METRICS:
        metric = "uploaded_count"

    af = build_access_filter(auth, "rv")
    sql = get_metric_query(metric, af)
    params = {"granularity": granularity, **af.params}

    rows = conn.execute(text(sql), params).mappings().all()
    points = [{"period": str(r["period"]), "value": float(r.get("value", 0))} for r in rows]

    latest = points[-1] if points else None
    previous = points[-2] if len(points) >= 2 else None
    delta_pct = None
    if latest and previous and previous["value"] != 0:
        delta_pct = round(((latest["value"] - previous["value"]) / previous["value"]) * 100, 2)

    return {
        "metric": metric,
        "granularity": granularity,
        "series": points,
        "summary": {
            "latestValue": round(latest["value"], 2) if latest else 0,
            "latestPeriod": latest["period"] if latest else None,
            "deltaVsPreviousPct": delta_pct,
        },
        "anomalies": get_trend_insights(points),
    }

File: frontend/backend/routes/usageTrends.js
const express = require('express');
const {
  METRIC_SQL,
  getTrendInsights,
  buildAccessFilter,
  getMetricQuery,
} = require('../queries/analyticsShared');

function createUsageTrendsRouter(pool) {
  const router = express.Router();

  router.get('/', async (req, res) => {
    const granularity = ['day', 'week', 'month', 'quarter'].includes(req.query.granularity)
      ? req.query.granularity
      : 'month';

    const metric = Object.prototype.hasOwnProperty.call(METRIC_SQL, req.query.metric)
      ? req.query.metric
      : 'uploaded_count';

    try {
      const accessFilter = buildAccessFilter(req.auth, 2, 'rv');
      const sql = getMetricQuery(metric, accessFilter);
      const { rows } = await pool.query(sql, [granularity, ...accessFilter.params]);
      const points = rows.map((r) => ({
        period: r.period,
        value:  Number(r.value || 0),
      }));

      const latest   = points[points.length - 1] || null;
      const previous = points[points.length - 2] || null;
      const deltaPct = latest && previous && previous.value !== 0
        ? ((latest.value - previous.value) / previous.value) * 100
        : null;

      res.json({
        metric,
        granularity,
        series: points,
        summary: {
          latestValue:         latest ? Number(latest.value.toFixed(2)) : 0,
          latestPeriod:        latest?.period || null,
          deltaVsPreviousPct:  deltaPct === null ? null : Number(deltaPct.toFixed(2)),
        },
        anomalies: getTrendInsights(points),
      });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

  return router;
}

module.exports = {
  createUsageTrendsRouter,
};
